aten/src/ATen/native/MaxUnpooling.cpp

```diff
@@ -0,0 +1,287 @@
+#include <tuple>
+#include "ATen/ATen.h"
+#include "ATen/NativeFunctions.h"
+
+namespace at {
+namespace native {
+
+template <typename scalar_t>
+at::Tensor MaxUnpooling2d_forward_out_cpu_(
+    Tensor& output,
+    const Tensor& input,
+    const Tensor& indices,
+    int64_t outputHeight,
+    int64_t outputWidth) {
+  // TODO: replicate is_empty() cbeck in SpatialMaxUnpooling.c
+  // TODO: zero out output inplace
+  auto numBatch = input.size(0);
+  auto numChannels = input.size(1);
+  auto inputHeight = input.size(2);
+  auto inputWidth = input.size(3);
+
+  auto* rawInput = input.data<scalar_t>();
+  auto* rawIndices = indices.data<int64_t>();
+  auto* rawOutput = output.data<scalar_t>();
+
+  int maxp;
+  for (auto n = 0; n < numBatch; n++) {
+    auto nOffset = n * numChannels * outputWidth * outputHeight;
+    for (auto k = 0; k < numChannels; k++) {
+      auto finalOffset = nOffset + k * outputWidth * outputHeight;
+      auto* output_p_k = rawOutput + finalOffset;
+      auto* input_p_k = rawInput + finalOffset;
+      auto* ind_p_k = rawIndices + finalOffset;
```

Comment:
This, and the line above, look wrong. The original code looks like this:

```
    scalar_t *output_p_k = output_p + k*owidth*oheight;
    scalar_t *input_p_k = input_p + k*iwidth*iheight;
    THIndex_t *ind_p_k = ind_p + k*iwidth*iheight;
```

You're adding the input and the ind by output area, when you should add input area.